### PR TITLE
feat(ui): abort promotion

### PR DESCRIPTION
**Pull request by Marvin9**
fixes https://github.com/akuity/kargo/issues/2662

https://github.com/user-attachments/assets/305e2d20-c2da-44a9-b4d0-6116798b5d67

### CODE REVIEW

File: internal/controller/promotions/promotions.go [hunk 1]
```diff
@@ -198,6 +198,7 @@ func newReconciler(
 	}
 	r.getStageFn = kargoapi.GetStage
 	r.promoteFn = r.promote
+	r.terminatePromotionFn = r.terminatePromotion
```
Comment (Marvin9) [inline internal/controller/promotions/promotions.go:201]: @hiddeco ^

Comment (hiddeco) [reply internal/controller/promotions/promotions.go:201]: Thanks 🙇

**Comment by Marvin9** [comment]
@hiddeco @krancour I think "Promotion terminated on <actor> request" message would be more useful

**Comment by hiddeco** [comment]
I don't think that in all cases the "actor" will be meaningful because it could be a controller identity. But happy to change it if @krancour agrees.

**Comment by Marvin9** [comment]
> it could be a controller identity

You mean controller aborts promotion itself in some cases?

**Comment by krancour** [comment]
> You mean controller aborts promotion itself in some cases?

When does it happen when not initiated by a user?

> I don't think that in all cases the "actor" will be meaningful because it could be a controller identity.

imho, if it's meaningful in _some_ cases, I think it's worth it.

**Comment by Marvin9** [comment]
Thanks @krancour @hiddeco this was an idea, and not blocking this PR

File: ui/src/features/common/promotion-directive-step-status/utils.ts [hunk 2]
```diff
@@ -30,7 +33,7 @@ export const getPromotionDirectiveStepStatus = (
 
   if (
     promotionStatus?.phase !== PromotionStatusPhase.SUCCEEDED &&
-    stepNumber > Number(promotionStatus?.currentStep)
+    stepNumber >= Number(promotionStatus?.currentStep)
```
Comment (Marvin9) [inline ui/src/features/common/promotion-directive-step-status/utils.ts:36]: @hiddeco can you confirm what is index of promotion step after it transition to Aborted state?

Comment (hiddeco) [reply ui/src/features/common/promotion-directive-step-status/utils.ts:36]: It will freeze.

Comment (Marvin9) [reply ui/src/features/common/promotion-directive-step-status/utils.ts:36]: Thanks for confirming. The change in this line is correct!

File: ui/src/features/stage/promotion-details-modal.tsx [hunk 3]
```diff
@@ -188,6 +234,22 @@ export const PromotionDetailsModal = ({
...
+            </Button>
+            <Button
+              danger
+              icon={<FontAwesomeIcon icon={faStopCircle} className='text-lg' />}
+              onClick={confirmAbortRequest}
+            >
+              Abort
+            </Button>
+          </>
+        )
```
Comment (rpelczar) [inline ui/src/features/stage/promotion-details-modal.tsx:251]: I think Close button should always be visible

File: ui/src/features/stage/promotion-details-modal.tsx [hunk 4]
```diff
@@ -188,6 +234,22 @@ export const PromotionDetailsModal = ({
...
+            <Button onClick={hide} type='primary'>
+              Close
+            </Button>
+            <Button
+              danger
+              icon={<FontAwesomeIcon icon={faStopCircle} className='text-lg' />}
+              onClick={confirmAbortRequest}
+            >
+              Abort
+            </Button>
```
Comment (rpelczar) [inline ui/src/features/stage/promotion-details-modal.tsx:249]: I wonder if this place is the best. On the right side, next to the cancel button usually we have confirmation/main action button. I like the danger color, but WDYT about moving this button to the left side?

File: ui/src/features/stage/promotion-details-modal.tsx [hunk 5]
```diff
@@ -188,6 +234,22 @@ export const PromotionDetailsModal = ({
       onOk={hide}
       onCancel={hide}
       cancelButtonProps={{ hidden: true }}
+      footer={
+        canAbortPromotion(promotion) && (
+          <>
+            <Button onClick={hide} type='primary'>
```
Comment (rpelczar) [inline ui/src/features/stage/promotion-details-modal.tsx:240]: Can we change the type to the default as everywhere for the cancel button?

File: ui/src/features/stage/promotions.tsx [hunk 6]
```diff
@@ -121,6 +142,11 @@ export const Promotions = () => {
       render: (_, promotion) => {
         const annotation = promotion.metadata?.annotations['kargo.akuity.io/create-actor'];
         const email = annotation ? annotation.split(':')[1] : 'N/A';
+
+        if (!email) {
+          return annotation;
+        }
+
         return email;
```
Comment (rpelczar) [inline ui/src/features/stage/promotions.tsx:150]:
```suggestion
        return email || annotation;
```

File: ui/src/features/stage/promotion-details-modal.tsx (hunk 5 above)
Comment (Marvin9) [reply ui/src/features/stage/promotion-details-modal.tsx]: Its default everywhere else so I changed to default here

File: ui/src/features/stage/promotion-details-modal.tsx (hunk 4 above)
Comment (Marvin9) [reply ui/src/features/stage/promotion-details-modal.tsx:247]: Makes sense but everywhere else we have action buttons from right to left. So if we change it here, we have to change at lots of places

File: ui/src/features/stage/promotion-details-modal.tsx (hunk 3 above)
Comment (Marvin9) [reply ui/src/features/stage/promotion-details-modal.tsx]: Done. Thanks

**Review by rpelczar — APPROVED** [review]

**Pull request merged**